File: prototypes/AI_Deltas/LKB/D7F3L9P2Q5R8/src/phasecube_delta/multigrid.py
"""Tri-grid orchestration with bias/delay feedback (SV6/SV7)."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Dict

import numpy as np

from .bias import DelayLine, make_bias_field
from .config import SimulationConfig
from .grid import PhaseGrid, GridMetrics
from .lenses import LensScheduler


@dataclass
class SwarmSnapshot:
    step: int
    lens: Dict[str, float]
    core: GridMetrics
    echo: GridMetrics
    memory: GridMetrics


class MultiGridSwarm:
    def __init__(self, config: SimulationConfig, seed: int = 0) -> None:
        self.config = config
        self.rng = np.random.default_rng(seed)
        self.core = PhaseGrid(config, self.rng, label="core")
        self.echo = PhaseGrid(config, self.rng, label="echo")
        self.memory = PhaseGrid(config, self.rng, label="memory")
        self.delay = DelayLine(length=config.delay_length, decay=config.delay_decay)
        self.scheduler = LensScheduler(config.lens_schedule)

    def step(self, step_idx: int, base_bias_mag: float) -> SwarmSnapshot:
        lens_weights = self.scheduler.active_weights(step_idx)

        # Influence-only bias fields
        base_bias = make_bias_field(self.config.grid_size, self.rng, base_bias_mag)
        self.delay.push(base_bias)
        delayed = self.delay.replay()

        # Cross-talk: echo/memory feed soft bias into core
        echo_bias = self.echo.liquid * self.config.coupling_echo_to_core
        memory_bias = self.memory.solid * self.config.coupling_memory_to_core

        core_bias = base_bias + delayed + echo_bias + memory_bias
        echo_self_bias = base_bias * 0.5 + self.core.liquid * 0.08
        memory_self_bias = delayed * 0.7 + base_bias * 0.3

        # Perturb before stepping to maintain non-collapse (SV2)
        self.core.perturb()
        self.echo.perturb()
        self.memory.perturb()

        core_metrics = self.core.step(core_bias, lens_weights)
        echo_metrics = self.echo.step(echo_self_bias, lens_weights)
        memory_metrics = self.memory.step(memory_self_bias, lens_weights)
        # TODO: add audio/file-fed bias injection once I/O is reintroduced; kept headless per scope.

        return SwarmSnapshot(
            step=step_idx,
            lens=lens_weights,
            core=core_metrics,
            echo=echo_metrics,
            memory=memory_metrics,
        )

    def run(self, steps: int, base_bias_mag: float) -> Dict[str, float]:
        snapshots = []
        for i in range(steps):
            snapshots.append(self.step(i, base_bias_mag))

        return self._summarize(snapshots)

    def _summarize(self, snapshots) -> Dict[str, float]:
        core_energy = np.mean([s.core.energy for s in snapshots]) if snapshots else 0.0
        echo_energy = np.mean([s.echo.energy for s in snapshots]) if snapshots else 0.0
        memory_energy = np.mean([s.memory.energy for s in snapshots]) if snapshots else 0.0
        core_div = np.mean([s.core.divergence for s in snapshots]) if snapshots else 0.0
        forgiveness_events = snapshots[-1].core.forgiveness_events if snapshots else 0
        return {
            "core_energy": float(core_energy),
            "echo_energy": float(echo_energy),
            "memory_energy": float(memory_energy),
            "core_divergence": float(core_div),
            "core_forgiveness_events": int(forgiveness_events),
            "steps": len(snapshots),
        }

File: prototypes/AI_Deltas/LKB/D3T4L7A9B12C/src/multiGrid.js
import { BiasField } from './biasField.js';
import { DelayLine } from './delayLine.js';
import { PhaseGrid } from './phaseGrid.js';

export class MultiGridSwarm {
  constructor(config, lensController, rng = Math) {
    this.config = config;
    this.lensController = lensController;
    this.biasField = new BiasField(config.gridSize, config.bias.decay);
    this.delayLine = new DelayLine(config.delay.length, config.delay.decay);
    this.core = new PhaseGrid(config.gridSize, config, rng);
    this.echo = new PhaseGrid(config.gridSize, config, rng);
    this.memory = new PhaseGrid(config.gridSize, config, rng);
    this.metrics = [];
  }

  injectPulse({ value = 0.2, radius = 1 } = {}) {
    const centerIndex = Math.floor(this.biasField.length / 2);
    this.biasField.injectSphere(centerIndex, radius, value);
  }

  tick(externalBiasValue = 0) {
    this.biasField.decay();
    if (externalBiasValue) {
      this.biasField.injectUniform(externalBiasValue);
    }

    this.delayLine.push(this.biasField.field);
    const delayBias = this.delayLine.mix();

    const lensWeights = this.lensController.tick();

    // Step echo and memory first to capture auxiliary influence.
    const echoMetrics = this._stepGrid(this.echo, delayBias, lensWeights);
    const memoryMetrics = this._stepGrid(this.memory, delayBias, lensWeights);

    // Cross-talk into bias field respecting coupling weights.
    for (let i = 0; i < this.biasField.length; i += 1) {
      this.biasField.field[i] += this.echo.liquid[i] * this.config.coupling.echoToCore;
      this.biasField.field[i] += this.memory.liquid[i] * this.config.coupling.memoryToCore;
    }

    const coreMetrics = this._stepGrid(this.core, delayBias, lensWeights);

    const snapshot = {
      lensWeights,
      core: coreMetrics,
      echo: echoMetrics,
      memory: memoryMetrics
    };
    this.metrics.push(snapshot);
    return snapshot;
  }

  _stepGrid(grid, delayBias, lensWeights) {
    grid.perturb();
    return grid.step(this.biasField, lensWeights, { delayBias });
  }
}
